From 5164357496c7b1a3ec62f8b4d2128efc063430b5 Mon Sep 17 00:00:00 2001
From: Tamal Saha <tasaha@habana.ai>
Date: Wed, 10 Jan 2024 11:22:44 +0200
Subject: [PATCH] [SW-161726] gaudi3: skip bist if cold rst bit set

This patch adds support to skip sram bist for 96MB cache
if cold rst bit (bist_skip_enable) is set or reset type
is FW_ALL.

FW issues hard reset after running bist. So to avoid rerun
of bist operation preboot sets bist_skip_enable bit in COLD
RESET register. Preboot checks this bit and ignores running
bist if it is set.

Ppboot skips bist operation if reset type is FW_ALL.

Change-Id: I16036e102e6b27a889173088388df71ec00d569a
Signed-off-by: Tamal Saha <tasaha@habana.ai>
---

diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
index b8a62e5..c3c53f2 100644
--- a/arch/gaudi3/arc/asic.c
+++ b/arch/gaudi3/arc/asic.c
@@ -105,10 +105,22 @@
 
 void hl_sram_self_repair(void)
 {
-
+	struct gaudi3_cold_rst_data cold_rst_data;
 	u64 sram_loadaddr;
 	bool ret;
 
+	/* If FW_ALL reset then skip bist */
+	if (hl_rst_src_get() == HL_FW_ALL_RST) {
+		preboot_puts("bist skipped for 96 MB SRAMs during FW all reset\n");
+		return;
+	}
+
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_DATA);
+	if (cold_rst_data.bist_skip_enable) {
+		preboot_puts("bist skipped for 96 MB SRAM\n");
+		return;
+	}
+
 	ret = ppboot_load_sram_mbist_data_img(&sram_loadaddr);
 	if (!ret) {
 		preboot_puts("SRAM MBIST image loading failed\n");
